contribs/dvrp/src/main/java/org/matsim/contrib/dvrp/passenger/GroupPassengerEngine.java

```diff
@@ -0,0 +1,231 @@
+		this.groupIdentifier = groupIdentifier;
+
+		internalPassengerHandling = new InternalPassengerHandling(mode, eventsManager);
+	}
+
+	@Override
+	public void setInternalInterface(InternalInterface internalInterface) {
+		this.internalInterface = internalInterface;
+		internalPassengerHandling.setInternalInterface(internalInterface);
+	}
+
+	@Override
+	public void onPrepareSim() {
+	}
+
+	@Override
+	public void doSimStep(double time) {
+		handleDepartures(time);
+		while (!rejectedRequestsEvents.isEmpty()) {
+			List<MobsimPassengerAgent> passengers = activePassengers.remove(rejectedRequestsEvents.poll().getRequestId());
+			//not much else can be done for immediate requests
+			//set the passenger agent to abort - the event will be thrown by the QSim
+			for (MobsimPassengerAgent passenger : passengers) {
+				passenger.setStateToAbort(mobsimTimer.getTimeOfDay());
+				internalInterface.arrangeNextAgentState(passenger);
+			}
+		}
+	}
+
+	@Override
+	public void afterSim() {
+	}
+
+	@Override
+	public boolean handleDeparture(double now, MobsimAgent agent, Id<Link> fromLinkId) {
+
+		if (!agent.getMode().equals(mode)) {
+			return false;
+		}
+
+		MobsimPassengerAgent passenger = (MobsimPassengerAgent)agent;
+		internalInterface.registerAdditionalAgentOnLink(passenger);
+		currentDepartures.add(passenger);
+		return true;
+	}
+
+	private void handleDepartures(double now) {
+		Map<Id<PassengerGroupIdentifier.PassengerGroup>, List<MobsimPassengerAgent>> agentGroups =
+				currentDepartures.stream().collect(
+						Collectors.groupingBy(
+								groupIdentifier,
+								Collectors.collectingAndThen(Collectors.toList(), list -> {
+									list.sort(Comparator.comparing(MobsimPassengerAgent::getId));
+									return list;
+								})));
+
+		for (List<MobsimPassengerAgent> group : agentGroups.values()) {
+
+			Iterator<MobsimPassengerAgent> iterator = group.iterator();
+			MobsimAgent firstAgent = iterator.next();
+			MobsimPassengerAgent passenger = (MobsimPassengerAgent) firstAgent;
+
+			Id<Link> toLinkId = firstAgent.getDestinationLinkId();
+
+			while (iterator.hasNext()) {
+				MobsimAgent next = iterator.next();
+				Gbl.assertIf(firstAgent.getCurrentLinkId().equals(next.getCurrentLinkId()));
+				Gbl.assertIf(toLinkId.equals(next.getDestinationLinkId()));
+			}
+
+			Route route = ((Leg)((PlanAgent)passenger).getCurrentPlanElement()).getRoute();
+
+			PassengerRequest request = requestCreator.createRequest(internalPassengerHandling.createRequestId(),
+					group.stream().map(Identifiable::getId).toList(), route,
+					getLink(firstAgent.getCurrentLinkId()), getLink(toLinkId),
+					now, now);
+
+
+			// must come before validateAndSubmitRequest (to come before rejection event)
+			eventsManager.processEvent(new PassengerWaitingEvent(now, mode, request.getId(), group.stream().map(Identifiable::getId).toList()));
+
+			validateAndSubmitRequest(List.copyOf(group), request, mobsimTimer.getTimeOfDay());
```

Comment:
Nit: No need to create an immutable copy when passing the list to a private method (inside your class).